tests/periph_gpio/main.c

```diff
@@ -19,74 +19,194 @@
+    else {
+        printf("PORT_%i.%02i is LOW\n", port, pin);
+    }
+    return 0;
+}
+
+static int set(int argc, char **argv)
+{
+    int port, pin;
 
-    puts("Now toggling pins 6 times with 500ms interval");
-    for (int c = 0; c < 6; c++) {
-        for (int i = 0; i < GPIO_NUMOF; i++) {
-            gpio_toggle(i);
-        }
-        hwtimer_wait(500000);
+    if (argc < 3) {
+        printf("usage: %s <port> <pin>\n", argv[0]);
+        return 1;
     }
+    port = atoi(argv[1]);
+    pin = atoi(argv[2]);
+
+    gpio_set(GPIO(port, pin));
+    return 0;
+}
 
+static int clear(int argc, char **argv)
+{
+    int port, pin;
 
-    puts("\n--------------------------------------------\n");
-    puts("Setting GPIOs to input w/o pull resister now");
-    for (int i = 0; i < GPIO_NUMOF; i++) {
-        gpio_init_in(i, GPIO_NOPULL);
+    if (argc < 3) {
+        printf("usage: %s <port> <pin>\n", argv[0]);
+        return 1;
     }
+    port = atoi(argv[1]);
+    pin = atoi(argv[2]);
+    gpio_clear(GPIO(port, pin));
+    return 0;
+}
 
-    puts("Will now poll all pins and print status on change");
+static int toggle(int argc, char **argv)
+{
+    int port, pin;
 
-    while (1) {
-        state = 0;
-        for (int i = 0; i < GPIO_NUMOF; i++) {
-            if (gpio_read(i)) {
-                state |= (1 << i);
-            }
-        }
-        if (state != old_state) {
-            for (int i = 0; i < GPIO_NUMOF; i++) {
-                if (state & (1 << i)) {
-                    printf("GPIO_%i:H  ", i);
-                } else {
-                    printf("GPIO_%i:L  ", i);
-                }
-            }
-            printf("\n");
-            old_state = state;
-        }
+    if (argc < 3) {
+        printf("usage: %s <port> <pin>\n", argv[0]);
+        return 1;
     }
+    port = atoi(argv[1]);
+    pin = atoi(argv[2]);
+    gpio_toggle(GPIO(port, pin));
+    return 0;
+}
+
+static const shell_command_t shell_commands[] = {
+    { "init_in", "initialize pin as output", init_in },
+    { "init_out", "initialize pin as input", init_out },
+    { "init_exti", "initialize pin as EXTI", init_exti },
+    { "read", "read pin status", read },
+    { "set", "set pin to HIGH", set },
+    { "clear", "set pin to LOW", clear },
+    { "toggle", "toggle pin", toggle },
+    { NULL, NULL, NULL }
+};
+
+int main(void)
+{
+    shell_t shell;
+
+    puts("GPIO peripheral driver test\n");
+    puts("In this test, pins are specified by integer port and pin numbers.\n"
+         "So if your platform has a pin PA01, it will be port=0 and pin=1,\n"
+         "PC14 would be port=2 and pin=14 etc.\n\n"
+         "NOTE: make sure the values you use exist on your platform! The\n"
+         "      behavior for not existing ports/pins is not defined!");
+
+    /* start the shell */
+    (void) posix_open(uart0_handler_pid, 0);
+    shell_init(&shell, shell_commands, SHELL_BUFSIZE, uart0_readc, uart0_putc);
+    shell_run(&shell);
 
     return 0;
 }
```

Comment:
I agree with Kaspar. But this test is of course not the end of it, there is room for improvement. I could think of something like defining a set of board specific GPIO pins that should be tested and providing some shell commands, that work on this set. But Lets take it step-by-step and merge this first and improve afterwards, otherwise this PR gets too heavy and it will never be merged...